VTK Mesa, Python3 build

milljm opened the pull request:
Build VTK with custom Mesa for off-screen rendering (all Civet test machines will be able to test ImageDiff tests)

Move towards a Python3 environment.

The package build steps _should_ pass, but all testing is going to fail, and continue to do so until we merge https://github.com/idaholab/moose/pull/13903

Commit:
Python3, MESA additions

Move towards a usable Python3.
Add MESA support for offscreen rendering (civet testing)

Commit:
Fix encodings, add more packages

Obtain latest QT and PyQT for Python3

Obtain python-config (python3-config) using pip (miniconda
symlinks pip to pip3)

Add additional civet specific modules (VTK-Mesa is not compiler
specific yay!)

milljm commented:
Ubuntu-16 can not be fixed until tomorrow morning. The fix requires a singularity container update. Doing so while the build machines are busy, is not possible.

Commit:
Use python3.7, force python2 for slepc

Commit:
Lets build from source, not a git repo

Instead of pulling from a repo, let us build from a pre-configured
autoconf routine. Fedora-30 has strange issues with autoconf...
'Mesa' is not the only project which does not './autogen.sh'.

Commit:
Print job errors correctly

Commit:
Build LLDB with Python3

Changes necessary to link with Python3 libraries when building
LLDB.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/391683/) on 5a90be4 : invalidated by @milljm

Trying to pull python-3 branch

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/391683/) on 5a90be4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/209/site)

This comment will be updated on new commits.

Commit:
Fix exception call

Convert exception call to Python3.

Commit:
Byte to string conversion fix

Commit:
Attempt Darwin build with Python3

Changes necessary to get vtk-mesa built on our Darwin targets.
First commit to see how far our Darwin targets succeed.

Commit:
Upgrade PETSc, GCC, LLVM

Upgrade PETSC_DEFAULT to 3.11, so we can bump ALT to a version that supports Python3.
Might as well attempt latest GCC/LLVM.

moosebuild commented:
Job [CentOS Package Builder](https://civet.inl.gov/job/407061/) on 54e9d00 : invalidated by @milljm

Built CentOS 7.7 container with fixed /usr/local/bin/python3

Commit:
Fix PETSC_ALT recipes

Looks like Sierra and High Sierra are just not ready to move to LLVM 9.
Forgot to bump linux GCC/LLVM versions.

milljm commented:
Looks like SLEPc 3.10.x also does not support python3. We are going to have to move PETSc ALT to 3.12 me thinks.

moosebuild commented:
Job [Sierra Package Builder](https://civet.inl.gov/job/407189/) on 915970f : invalidated by @milljm

Python2 is available, just not linked properly.

moosebuild commented:
Job [Ubuntu 16 Package Builder](https://civet.inl.gov/job/407190/) on 915970f : invalidated by @milljm

Fixed Ubuntu-16 container. Ubuntu 16 uses python3.5, which has an issue with thread_pool (apparently)

Commit:
Fix more Python2-3 issues

Turns out mumps is failing with missing libraries (mpiseq). Not sure why this
is just now cropping up. The only difference between the last successful build
is Python3 (and I suppose GCC8 to 9. But the failures are with the LLVM compiled
libraries, and that has not changed).

Removing the call to python2. Use environment defaults instead.

Commit:
Use SYSROOT for Darwin Trilinos builds

Commit:
Things necessary to get Catalina in play

Hopefully these changes also work for pre-catalina builds. I don't why not. We should have been using sysroot a long time ago.

Commit:
Revert "Things necessary to get Catalina in play"

This reverts commit 05ebd3a481f254a8720a275ead623e1cd5505ffc.

Nope. Not the right solution.

Commit:
Xcode 11 isysroot LLVM

Remove unnecessarily complicated conditionals.
Add catalina support.
Fix recent issue with conda update conda dropping setuptools.

Commit:
Normalize petsc configure lines, fix python-config

When you install Miniconda3, you get a symlink from python to python3.
But not a symlink from python-config to python3-config.

Commit:
Remove unnecessary config lines to LLVM

Also, trigger the new Catalina target. Expect some issues. LLDB, Certs, ssh keys, etc.

Commit:
Use pkgbuild/productbuild

PackageMaker is no longer viable (32bit).
Fix modules and zsh compatibilities
Update Doxygen.
Update CMake.

Commit:
Modify path to modules init

Commit:
Update python library LLDB links with

With the change to Python3, the library path changed slightly.

Commit:
Several libraries/binaries linking to wrong python

Finding more python libraries lldb is linking incorrectly.

This is still a work in progress, as it seems lldb-server is not
launching correctly.

Commit:
Remove deprecated patch, silence modules

Doxygen patch in place was for different version.

Silence the module loading messages. It would seem civet is treating
any returned text as errors.

Commit:
Fix Darwin LLDB, add gperftools to default modules

Instead of downloading individual llvm projects, use the rather
convenient git repo. This method also seems to remove the need
to build lldb separately.

Add gperftools to the default modules Refs https://github.com/idaholab/moose/pull/14212

Commit:
Fix download links

Commit:
LLVM was being built in debug mode maybe?

Everything being built by LLVM on darwin, is extremely slow. I think
everything is being built without any optimizations (not release).

Commit:
Do LLVM post_run properly

Do code-signing.
Create clang python bindings.

Commit:
I can not codesign lldb-8.0.1. Update to 9.

There is something wrong with trying to codesign LLDB on 8.0.x.
Lets upgrade to 9, as that appears to work on my workstation.

LLVM _still_ links to rapth Python, which in the end, is not the
python library I am asking LLDB to link to. It is still necessary
to use install_name_tool to fix these issues.

Use the correct codesign certificate path in the Keychain.

Commit:
Add VTK to conda install list

Conda's VTK seems to work (we no longer need idaholab channel).
Remove old remnants of VTK6 environment variables.

Commit:
Add mumps back to PETSc contribs

We removed mumps previously, to resolve _some_ forgotten issue.
Attempt a build with mumps added again.

milljm commented:
@fdkong, looks like adding mumps is not possible. The following is happening on all builds (Catalina, Ubuntu, CentOS... etc). It is probably either not supposed to be enabled for Debug versions of PETSc, or perhaps Python3 related? I doubt python has anything to do with the installation of mumps. The following error says it failed to _install_ mumps, not build it.

Unfortunately I do not have the log file. But I will attempt to build this manually and report.

```pre
./configure --prefix=/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-dbg \
--with-debugging=1 \
--with-ssl=0 \
--with-pic=1 \
--with-openmp=1 \
--with-shared-libraries=1 \
--with-cxx-dialect=C++11 \
--download-hypre=1 \
--download-fblaslapack=1 \
--download-metis=1 \
--download-ptscotch=1 \
--download-parmetis=1 \
--download-superlu_dist=1 \
--download-scalapack=1 \
--download-mumps=1
====

===============================================================================
    Compiling PTScotch; this may take several minutes
===============================================================================
===============================================================================
    Installing PTScotch; this may take several minutes
===============================================================================
===============================================================================
    Trying to download git://https://bitbucket.org/petsc/pkg-metis.git for METIS
===============================================================================
===============================================================================
    Configuring METIS with cmake, this may take several minutes
===============================================================================
===============================================================================
    Compiling and installing METIS; this may take several minutes
===============================================================================
===============================================================================
    Trying to download git://https://bitbucket.org/petsc/pkg-parmetis.git for PARMETIS
===============================================================================
===============================================================================
    Configuring PARMETIS with cmake, this may take several minutes
===============================================================================
===============================================================================
    Compiling and installing PARMETIS; this may take several minutes
===============================================================================
===============================================================================
    Trying to download http://ftp.mcs.anl.gov/pub/petsc/externalpackages/fblaslapack-3.4.2.tar.gz for FBLASLAPACK
===============================================================================
===============================================================================
    Compiling FBLASLAPACK; this may take several minutes
===============================================================================
===============================================================================
    Trying to download git://https://github.com/xiaoyeli/superlu_dist for SUPERLU_DIST
===============================================================================
===============================================================================
    Configuring SUPERLU_DIST with cmake, this may take several minutes
===============================================================================
===============================================================================
    Compiling and installing SUPERLU_DIST; this may take several minutes
===============================================================================
===============================================================================
    Trying to download git://https://bitbucket.org/petsc/pkg-scalapack for SCALAPACK
===============================================================================
===============================================================================
    Compiling and installing Scalapack; this may take several minutes
===============================================================================
===============================================================================
    Trying to download git://https://bitbucket.org/petsc/pkg-mumps.git for MUMPS
===============================================================================
===============================================================================
    Compiling Mumps; this may take several minutes
===============================================================================
===============================================================================
    Installing Mumps; this may take several minutes
===============================================================================
*******************************************************************************
         UNABLE to CONFIGURE with GIVEN OPTIONS    (see configure.log for details):
-------------------------------------------------------------------------------
Downloaded mumps could not be used. Please check install in /opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-dbg
*******************************************************************************
```

Commit:
Remove MUMPS for clang builds

It would appear, my clang builds are failing to build PETSc when also
asking PETSc to build with MUMPS. Disable for now.

Fix the codesign logic when not codesigning LLDB.

milljm commented:
What is failing has been fixed (or will be) in moose/next. Time to merge, so we can have a python-3 environment available in Civet.
